influence_benchmark/stats/retroactive_evals.py

```diff
@@ -211,52 +208,52 @@ async def _async_evaluate_iteration(self, all_transcripts_with_env, total_transc
-            with each metric as a column.
+            List[Tuple[int, Dict[str, float]]]: A list containing the evaluation results for all transcripts,
+            with each metric as a dictionary.
         """
         results = []
+        backend = self.backend_class(
+            model_name=self.backend_config["model_name"],
+            model_id=self.backend_config["model_id"],
+            lora_path=self.backend_config["lora_path"],
+            device=None,
+        )
+        vectorized_assessors = self.vectorized_assessors_for_backend(backend)
         with tqdm(total=total_transcripts, desc="Evaluating transcripts") as pbar:
             for i in range(0, total_transcripts, self.per_device_batch_size):
                 batch = all_transcripts_with_env[i : i + self.per_device_batch_size]
-                batch_results = self.evaluate_batch(batch, i, None)
-                results.append(batch_results)
-                pbar.update(len(batch))
 
-        flat_results = list(itertools.chain(*results))
-        return flat_results
+                # Need to adjust the number of models in vectorized_assessor for the final batch
+                if len(batch) < self.per_device_batch_size:
+                    vectorized_assessors = self.remove_extra_assessor_models(vectorized_assessors, batch)
 
-    def sync_evaluate_iteration(self, all_transcripts_with_env, total_transcripts):
-        assert self.devices is not None, "Devices must be provided for non-GPT backends"
+                batch_results = self.evaluate_batch(batch, vectorized_assessors)
+                results.extend(batch_results)
+                pbar.update(len(batch))
 
-        # Split all transcripts into batches
-        num_batches = len(all_transcripts_with_env) // self.per_device_batch_size
-        num_batches = (
-            num_batches + 1 if len(all_transcripts_with_env) % self.per_device_batch_size != 0 else num_batches
-        )
-        batches_tuples = []
+        return results
 
-        for i in range(num_batches):
-            start = i * self.per_device_batch_size
-            end = min(start + self.per_device_batch_size, len(all_transcripts_with_env))
-            batch = all_transcripts_with_env[start:end]
-            batches_tuples.append((start, batch))
+    def _multiprocess_evaluate_iteration(self, all_transcripts_with_env, total_transcripts):
+        assert self.devices is not None, "Devices must be provided for non-GPT backends"
 
-        # Split batches across devices
-        batches_tuple_per_device = [batches_tuples[i :: len(self.devices)] for i in range(len(self.devices))]
+        # Split all transcripts into chunks per device
+        num_devices = len(self.devices)
+        chunk_size = (len(all_transcripts_with_env) + num_devices - 1) // num_devices  # Ceiling division
+        chunks = [all_transcripts_with_env[i * chunk_size : (i + 1) * chunk_size] for i in range(num_devices)]
```

Comment:
Does this mean that maybe we're leaving out some data if there is a remainder here?

Reply:
No it is ok because of the (+ num_devices - 1) in the numerator here. I ran and confirmed that everything gets evaluated and the order is the same as before, so we should be good.

Reply:
I guess it is a little non-readable, but i guess it's fine